identity-did/src/verifiable/document_verifier.rs

```diff
@@ -0,0 +1,230 @@
+    self
+  }
+
+  /// Verify the signing verification method type matches one specified.
+  pub fn method_type(mut self, method_type: Vec<MethodType>) -> Self {
+    self.options = self.options.method_type(method_type);
+    self
+  }
+
+  /// Verify the [`Signature::challenge`] field matches this.
+  pub fn challenge(mut self, challenge: String) -> Self {
+    self.options = self.options.challenge(challenge);
+    self
+  }
+
+  /// Verify the [`Signature::domain`] field matches this.
+  pub fn domain(mut self, domain: String) -> Self {
+    self.options = self.options.domain(domain);
+    self
+  }
+
+  /// Verify the [`Signature::purpose`] field matches this. Also verifies that the signing
+  /// method has the corresponding verification method relationship.
+  ///
+  /// E.g. [`ProofPurpose::Authentication`] must be signed using a method
+  /// with [`MethodRelationship::Authentication`].
+  ///
+  /// NOTE: `purpose` overrides the `method_scope` option.
+  pub fn purpose(mut self, purpose: ProofPurpose) -> Self {
+    self.options = self.options.purpose(purpose);
+    self
+  }
+
+  /// Determines whether to error if the current time exceeds the [`Signature::expires`] field.
+  ///
+  /// Default: false (reject expired signatures).
+  pub fn allow_expired(mut self, allow_expired: bool) -> Self {
+    self.options = self.options.allow_expired(allow_expired);
+    self
+  }
+}
+
+impl<T, U, V> DocumentVerifier<'_, T, U, V>
+where
+  U: Revocation,
+{
+  /// Verifies the signature of the provided data.
+  ///
+  /// # Errors
+  ///
+  /// Fails if an unsupported verification method is used, data
+  /// serialization fails, or the verification operation fails.
+  pub fn verify<X>(&self, data: &X) -> Result<()>
+  where
+    X: Serialize + TrySignature,
+  {
+    let signature: &Signature = data
+      .try_signature()
+      .map_err(|_| Error::InvalidSignature("missing signature"))?;
+
+    // Retrieve the method used to create the signature and check it has the required verification
+    // method relationship (purpose takes precedence over method_scope).
+    let purpose_scope = self.options.purpose.map(|purpose| match purpose {
+      ProofPurpose::AssertionMethod => MethodScope::assertion_method(),
+      ProofPurpose::Authentication => MethodScope::authentication(),
+    });
+    let method: &VerificationMethod<U> = if let Some(purpose_scope) = purpose_scope {
+      self
+        .document
+        .try_resolve_method_with_scope(signature, purpose_scope)
+        .map_err(|_| Error::InvalidSignature("method with purpose scope not found"))?
+    } else if let Some(scope) = self.options.method_scope {
+      self
+        .document
+        .try_resolve_method_with_scope(signature, scope)
+        .map_err(|_| Error::InvalidSignature("method with specified scope not found"))?
+    } else {
+      self
+        .document
+        .try_resolve_method(signature)
+        .map_err(|_| Error::InvalidSignature("method not found"))?
+    };
```

Comment:
I find this more readable:
```suggestion
let method: &VerificationMethod<U> = match (purpose_scope, self.options.method_scope) {
      (Some(purpose_scope), _) => self
        .document
        .try_resolve_method_with_scope(signature, purpose_scope)
        .map_err(|_| Error::InvalidSignature("method with purpose scope not found"))?,
      (None, Some(scope)) => self
        .document
        .try_resolve_method_with_scope(signature, scope)
        .map_err(|_| Error::InvalidSignature("method with specified scope not found"))?,
      _ => self
        .document
        .try_resolve_method(signature)
        .map_err(|_| Error::InvalidSignature("method not found"))?,
    };
```